Add mortgage function, change property dataclass to include ID and l…

mulliganw opened the pull request:
…evel

Currently, property ID is loosely defined in the Game class and referenced in the Player dataclass. This PR attempts to streamline property ID and at least give it some weight. Constant property IDs are required for house building and mortgaging, because the player needs to access them from anywhere on the board.

mulliganw committed:
Add mortgage function, change property dataclass to include ID and level

Currently, property ID is loosely defined in the Game class and referenced in the Player dataclass. This PR attempts to streamline property ID and at least give it some weight. Constant property IDs are required for house building and mortgaging, because the player needs to access them from anywhere on the board.

mulliganw commented:
If we do go with this approach, the Company dataclass would also have to be updated to include id, but that pull isn't accepted at the time of this comment.

JackChuck1 commented:
I think instead of making the ID a bigger part of the code I think we should remove it entirely. We already have an array full of all the properties, the grid. The properties array seems redundant when we could just access grid[player.position]

TankTheTortoise commented:
> I think instead of making the ID a bigger part of the code I think we should remove it entirely. We already have an array full of all the properties, the grid. The properties array seems redundant when we could just access grid[player.position]

If we need to mortgage or trade a property, we won't be on the tile, so we need an id

JackChuck1 commented:
We can access that through the player array for their owned properties, I'm fine with having that array, just the one in the Game class seems unnecessary

JackChuck1 commented on monopoly.py:
```diff
@@ -72,6 +75,14 @@ class Player:
+        # Make sure the player owns the property they're attempting to mortgage and that it has no houses
+        if property in self.properties and property.get('level') == 0 :
+            # Remove the property from the player's list of properties
+            self.properties.remove(property.get('id'))
```
Mortgaged properties are still owned by the player they just don't give rent

mulliganw replied:
fixed

mulliganw commented:
> We can access that through the player array for their owned properties, I'm fine with having that array, just the one in the Game class seems unnecessary

Having ID allows more flexibility. if we want to make any changes to the original game of Monopoly, it would be easier for people to understand a straight ID system rather than having to check whose property it is and what tile it's located on and get its name from there and so on and so on. I think even if its redundant and less efficient it significantly increases the readability and saves a lot of time. 

tldr; 
just cuz it works now doesnt mean its gonna be convenient to access later

mulliganw committed:
Fix mortgaging so the player keeps the property

pretty sure it works but there's probably a better way to do it with .replace() or something

JackChuck1 commented:
I think we should pull this just without that one line with the id. I'm on board with the ID system now but the property IDs are just kinda what index they're at in the property array.

mulliganw committed:
remove id from property
